src/main/java/org/rascalmpl/maven/AbstractRascalMojo.java

```diff
@@ -0,0 +1,380 @@
+	protected MavenProject project;
+
+	@Parameter(defaultValue = "${project.build.outputDirectory}", property = "bin", required = true )
+	protected String bin;
+
+	@Parameter(defaultValue = "${project.build.directory}/generatedSources", property = "generatedSources", required = true)
+	protected String generatedSources;
+
+	@Parameter(property = "srcs", required = true )
+	protected List<String> srcs;
+
+	@Parameter(property = "srcIgnores", required = false )
+	protected List<String> srcIgnores;
+
+	@Parameter(property = "libs", required = false )
+	protected List<String> libs;
+
+	@Parameter(defaultValue="false", property= "verbose", required=true)
+	protected boolean verbose;
+
+	@Parameter(defaultValue = "${session}", required = true, readonly = true)
+  	protected MavenSession session;
+
+	@Parameter(defaultValue = "0.41.0-RC18", required = false, readonly = true)
+	protected String bootstrapRascalVersion;
+
+	@Parameter
+	private BuildPluginManager pluginManager;
+
+	@Parameter(defaultValue="")
+	protected String mainModule;
+
+	/**
+	 * Which `-Drascal.skipTag.skip` to use
+	 */
+	protected final String skipTag;
+
+	/**
+	 * Class with `static int main(String[] arg)` method to call
+	 */
+	protected final String mainClass;
+
+	/**
+	 * If there are more than the normal PathConfig parameters
+	 * they can be added here.
+	 */
+	protected Map<String, String> extraParameters = new HashMap<>();
+
+	/**
+	 * The Rascal runtime jar that will be used to load classes (including the main class) from.
+	 * This is where bootstrap issues are resolved.
+	 */
+	protected Path cachedRascalRuntime = null;
+
+	public AbstractRascalMojo(String mainClass, String skipTag) {
+		this.mainClass = mainClass;
+		this.skipTag = skipTag;
+	}
+
+	protected Path getRascalRuntime() {
+		if (cachedRascalRuntime != null) {
+			cachedRascalRuntime = detectedDependentRascalArtifact(getLog(), project);
+		}
+
+		return cachedRascalRuntime;
+	}
+
+	protected List<Path> locations(List<String> paths) {
+		return paths.stream()
+			.map(Path::of)
+			.collect(Collectors.toCollection(ArrayList::new));
+	}
+
+	@Override
+	public void execute() throws MojoExecutionException {
+		try {
+			if (System.getProperty("rascal." + skipTag + ".skip") != null) {
+				getLog().info("Skipping " + getClass().getName() + " completely");
+				return;
+			}
+
+			Path binLoc = Path.of(bin);
+			Path generatedSourcesLoc = Path.of(generatedSources);
+			List<Path> srcLocs = locations(srcs);
+			List<Path> ignoredLocs = locations(srcIgnores);
+			List<Path> libLocs = locations(libs);
+
+			getLog().info("configuring paths");
+			for (Path src : srcLocs) {
+				getLog().info("\tregistered source location: " + src);
+			}
+
+			for (Path ignore : ignoredLocs) {
+				getLog().warn("\tignoring sources in: " + ignore);
+			}
+
+			getLog().info("Checking if any files need compilation...");
+
+			List<Path> todoList = getTodoList(binLoc, srcLocs, ignoredLocs);
```

Comment:
is this always needed? like for the `rascal:shell`?

Reply:
No then you just override the execute method.